include/scrimmage/common/ParameterServer.h

```diff
@@ -0,0 +1,138 @@
+ *   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
+ *   License for more details.
+ *
+ *   You should have received a copy of the GNU Lesser General Public License
+ *   along with SCRIMMAGE.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @author Kevin DeMarco <kevin.demarco@gtri.gatech.edu>
+ * @author Eric Squires <eric.squires@gtri.gatech.edu>
+ * @date 31 July 2017
+ * @version 0.1.0
+ * @brief Brief file description.
+ * @section DESCRIPTION
+ * A Long description goes here.
+ *
+ */
+
+#ifndef INCLUDE_SCRIMMAGE_COMMON_PARAMETERSERVER_H_
+#define INCLUDE_SCRIMMAGE_COMMON_PARAMETERSERVER_H_
+
+#include <scrimmage/common/Parameter.h>
+
+#include <string>
+#include <map>
+#include <list>
+#include <memory>
+#include <iostream>
+#include <algorithm>
+
+class Plugin;
+using PluginPtr = std::shared_ptr<Plugin>;
+
+namespace scrimmage {
+class ParameterServer {
+ public:
+    void unregister_params(PluginPtr owner);
+
+    template <class T>
+    bool register_param(const std::string &name, T &variable,
+                        std::function<void(const T &value)> callback,
+                        PluginPtr owner) {
+        // Create the Parameter
+        auto param = std::make_shared<Parameter<T>>(variable, callback, owner);
+
+        // Search for the parameter name in the server
+        auto it_name = params_.find(name);
+        if (it_name == params_.end()) {
+            // The parameter name doesn't exist it, create it.
+            params_[name][typeid(T).name()] = std::list<ParameterBasePtr> {param};
+        } else {
+            // Does this type exist for this parameter name?
+            auto it_type = it_name->second.find(typeid(T).name());
+            if (it_type != it_name->second.end()) {
+                // This param name / type combo exists
+                // Has this specific plugin already registered this parameter?
+                for (const ParameterBasePtr &ptr : it_type->second) {
+                    if (ptr->owner() == owner) {
+                        return false;
+                    }
+                }
+            }
+            // Add the parameter to the params_ map
+            (it_name->second)[typeid(T).name()].push_back(param);
+        }
+        return true;
+    }
+
+    template <class T>
+    bool set_param(const std::string &name, const T &value) {
+        // Search for the parameter name in the server
+        auto it_name = params_.find(name);
+        if (it_name != params_.end()) {
+            // The param name exists, does the type exist?
+            auto it_type = it_name->second.find(typeid(T).name());
+            if (it_type != it_name->second.end()) {
+                // The param name/type exists in the list, iterate over the
+                // list of parameters and call their set_value functions
+                for (ParameterBasePtr param : it_type->second) {
+                    auto param_cast = std::dynamic_pointer_cast<Parameter<T>>(param);
```

Comment:
This seems inefficient to check every time ``set_param`` is called. I think we can move this type check to ``register_param`` so it is only done once.

Reply:
I thought about this a bit more and realized we do need the dynamic cast so please disregard.

Reply:
You are right. I wrote this check before I changed params_ to use have a second map that is keyed based on a string representation of the type (typeid(T).name()). This guarantees that the dynamic_pointer_cast will succeed later when called.